File: jsrc/legacy/SxxMachine/builtin/PRED_$read_token0_3.java
package SxxMachine;

import java.io.PushbackReader;

import SxxMachine.DoubleTerm;
import SxxMachine.FFIObjectTerm;
import SxxMachine.IntegerTerm;
import SxxMachine.LongTerm;
import SxxMachine.Operation;
import SxxMachine.Predicate;
import SxxMachine.Prolog;
import SxxMachine.SymbolTerm;
import SxxMachine.Term;
import SxxMachine.TermData;
import SxxMachine.Token;
import SxxMachine.VariableTerm;
import SxxMachine.ExistenceException;
import SxxMachine.IllegalDomainException;
import SxxMachine.JavaException;
import SxxMachine.PInstantiationException;
import SxxMachine.PermissionException;
/**
 * <code>'$read_token0'/3</code><br>
 * @author Mutsunori Banbara (banbara@kobe-u.ac.jp)
 * @author Naoyuki Tamura (tamura@kobe-u.ac.jp)
 * @version 1.0
 * @since 0.7
 */
class PRED_$read_token0_3 extends Predicate.P3 {
    public PRED_$read_token0_3(Term a1, Term a2, Term a3, Operation cont){
	LARG[0] = a1;
	LARG[1] = a2;
	LARG[2] = a3;
	this.cont = cont;
    }

    /* The a1 must be user, user_input, and
       java.io.PushbackReader, otherwise fails. */
    public Operation exec(Prolog engine) {
        engine.setB0();
	Term a1, a2, a3;
	a1 = LARG[0];
	a2 = LARG[1];
	a3 = LARG[2];

	Object stream = null;
	StringBuilder s;
	int type;
	Term token;

	// S_or_a
	a1 = a1.dref();
	if ((a1 instanceof VariableTerm)) {
	    throw new PInstantiationException(this, 1);
	} else if ((a1 instanceof SymbolTerm)) {
	    if (! engine.getStreamManager().containsKey(a1))
		throw new ExistenceException(this, 1, "stream", a1, "");
	    stream = ( engine.getStreamManager().get(a1)).object();
	} else if ((a1 instanceof FFIObjectTerm)) {
	    stream = a1.object();
	} else {
	    throw new IllegalDomainException(this, 1, "stream_or_alias", a1);
	}
	if (! (stream instanceof PushbackReader))
	    throw new PermissionException(this, "input", "stream", a1, "");
	// read token
	s = new StringBuilder();
	try {
	    type = Token.read_token(s, (PushbackReader)stream);
	    switch(type) {
	    case Token.TOKEN_INTEGER:
	    	token = new IntegerTerm(Integer.parseInt(s.toString()));
	    	break;
	    case Token.TOKEN_LONG:
	    	token = new LongTerm(Long.parseLong(s.toString()));
	    	break;
	    case Token.TOKEN_DOUBLE:
	    	token = new DoubleTerm(Double.parseDouble(s.toString()));
	    	break;
	    case 'S':
			char[] chars = (s.toString()).toCharArray();
			token = Prolog.Nil;
			for (int i=chars.length; i>0; i--){
			    token = TermData.CONS(new IntegerTerm((int)chars[i-1]), token);
			}
			break;
	    default :
	    	token = SymbolTerm.create(s.toString());
	    	break;
	    }
	} catch (Exception e) {
	    throw new JavaException(this, 1, e);
	}
	if (! a2.unify(new IntegerTerm(type), engine.trail))
	    return engine.fail();
	if (! a3.unify(token, engine.trail))
	    return engine.fail();
	return cont;
    }
}

File: src/legacy/SxxMachine/builtin/PRED_$read_token0_3.cpp
using namespace std;

#include "PRED_$read_token0_3.h"
#include "../../../machine/SxxMachine/Operation.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/Prolog.h"
#include "../../../machine/SxxMachine/VariableTerm.h"
#include "../../../exceptions/SxxMachine/PInstantiationException.h"
#include "../../../machine/SxxMachine/SymbolTerm.h"
#include "../../../exceptions/SxxMachine/ExistenceException.h"
#include "../../../machine/SxxMachine/FFIObjectTerm.h"
#include "../../../exceptions/SxxMachine/IllegalDomainException.h"
#include "../../../exceptions/SxxMachine/PermissionException.h"
#include "../../../machine/SxxMachine/Token.h"
#include "../../../machine/SxxMachine/IntegerTerm.h"
#include "../../../machine/SxxMachine/LongTerm.h"
#include "../../../machine/SxxMachine/DoubleTerm.h"
#include "../../../machine/SxxMachine/TermData.h"
#include "../../../exceptions/SxxMachine/JavaException.h"

namespace SxxMachine
{
	using DoubleTerm = SxxMachine::DoubleTerm;
	using FFIObjectTerm = SxxMachine::FFIObjectTerm;
	using IntegerTerm = SxxMachine::IntegerTerm;
	using LongTerm = SxxMachine::LongTerm;
	using Operation = SxxMachine::Operation;
	using Predicate = SxxMachine::Predicate;
	using Prolog = SxxMachine::Prolog;
	using SymbolTerm = SxxMachine::SymbolTerm;
	using Term = SxxMachine::Term;
	using TermData = SxxMachine::TermData;
	using Token = SxxMachine::Token;
	using VariableTerm = SxxMachine::VariableTerm;
	using ExistenceException = SxxMachine::ExistenceException;
	using IllegalDomainException = SxxMachine::IllegalDomainException;
	using JavaException = SxxMachine::JavaException;
	using PInstantiationException = SxxMachine::PInstantiationException;
	using PermissionException = SxxMachine::PermissionException;

	PRED_$read_token0_3::PRED_$read_token0_3(Term *a1, Term *a2, Term *a3, Operation cont)
	{
	LARG[0] = a1;
	LARG[1] = a2;
	LARG[2] = a3;
	this->cont = cont;
	}

	Operation PRED_$read_token0_3::exec(Prolog *engine)
	{
		engine->setB0();
	Term *a1, *a2, *a3;
	a1 = LARG[0];
	a2 = LARG[1];
	a3 = LARG[2];

	any stream = nullptr;
	StringBuilder *s;
	int type;
	Term *token;

	// S_or_a
	a1 = a1->dref();
	if ((dynamic_cast<VariableTerm*>(a1) != nullptr))
	{
		throw PInstantiationException(this, 1);
	}
	else if ((dynamic_cast<SymbolTerm*>(a1) != nullptr))
	{
		if (!engine->getStreamManager()->containsKey(a1))
		{
		throw ExistenceException(this, 1, "stream", a1, "");
		}
		stream = (engine->getStreamManager()->get(a1)).object();
	}
	else if ((dynamic_cast<FFIObjectTerm*>(a1) != nullptr))
	{
		stream = a1->object();
	}
	else
	{
		throw IllegalDomainException(this, 1, "stream_or_alias", a1);
	}
	if (!(dynamic_cast<PushbackReader*>(stream) != nullptr))
	{
		throw PermissionException(this, "input", "stream", a1, "");
	}
	// read token
	s = new StringBuilder();
	try
	{
		type = Token::read_token(s, any_cast<PushbackReader*>(stream));
		switch (type)
		{
		case Token::TOKEN_INTEGER:
			token = new IntegerTerm(static_cast<Integer>(s->toString()));
			break;
		case Token::TOKEN_LONG:
			token = new LongTerm(static_cast<long long>(s->toString()));
			break;
		case Token::TOKEN_DOUBLE:
			token = new DoubleTerm(static_cast<Double>(s->toString()));
			break;
		case 'S':
		{
			std::vector<char> chars = (s->toString()).toCharArray();
			token = Prolog::Nil;
			for (int i = chars.size(); i > 0; i--)
			{
				IntegerTerm tempVar(static_cast<int>(chars[i - 1]));
				token = TermData::CONS(&tempVar, token);
			}
			break;
		}
		default :
			token = SymbolTerm::create(s->toString());
			break;
		}
	}
	catch (const runtime_error &e)
	{
		throw JavaException(this, 1, e);
	}
	IntegerTerm tempVar2(type);
	if (!a2->unify(&tempVar2, engine->trail))
	{
		return engine->fail();
	}
	if (!a3->unify(token, engine->trail))
	{
		return engine->fail();
	}
	return cont;
	}
}
